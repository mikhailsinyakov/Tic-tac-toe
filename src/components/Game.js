import React from 'react';
import GameBoard from '@lib/GameBoard';
import Info from '@components/Info';
import Field from '@components/Field';
import Cell from '@components/Cell';

export default class Game extends React.Component {
    constructor(props) {
        super(props);

        this.fieldSide = 390;
        this.cellsInField = [3, 3];
        this.matchesToWin = 3;

        this.state = {
            board: new GameBoard(this.cellsInField[1], this.cellsInField[0]),
            turn: null
        };

        this.makeTurn = this.makeTurn.bind(this);
    }

    initField() {
        this.setState({
            turn: this.props.userSymbol == '×' ? 'user' : 'comp',
            board: new GameBoard(this.cellsInField[1], this.cellsInField[0])
        });
    }

    makeCompTurn() {
        const { state: { board }, props: { opponentSymbol, userSymbol }, matchesToWin } = this;
        const [ column, row ] = board.findBestMove(matchesToWin, opponentSymbol, userSymbol);

        // Simulate delay
        setTimeout(() => this.makeTurn(row, column, opponentSymbol), 1000);
    }

    makeTurn(row, column, symbol) {
        const symbols = ['×', '○'];
        if (!symbols.includes(symbol)) throw new Error('You should use either "×" or "○" symbol');
        
        const { state: { turn }, props: { stopGame, userSymbol, opponentSymbol, matchesToWin } } = this;

        const board = GameBoard.copy(this.state.board);
        board.addValue(row, column, symbol);
        this.setState({ board });
        
        const gameState = board.defineGameState(matchesToWin);
        if (gameState.state == 'in process') {
            this.setState({ turn: turn == 'user' ? 'comp' : 'user' });
        } else {
            let result;
            if (gameState.result == 'draw') result = 'Ничья';
            if (gameState.winner == userSymbol) result = 'Вы выиграли!';
            if (gameState.winner == opponentSymbol) result = 'Вы проиграли :(';
            this.setState({ turn: null });
            stopGame(result);
        }
    }

    componentDidUpdate(prevProps, prevState) {
        const gameStartPlaying = !prevProps.isPlaying && this.props.isPlaying;
        const compGotTurn = prevState.turn != 'comp' && this.state.turn == 'comp';

        if (gameStartPlaying) this.initField();
        if (compGotTurn) this.makeCompTurn();
    }

    render() {
        const { makeTurn, state: { board, turn }, props: { userSymbol, isPlaying } } = this;
        const [ columns ] = this.cellsInField;

        const cells = [];
        
        board.getValues().forEach((row, rowIndex) => {
            row.forEach((val, columnIndex) => {
                cells.push(<Cell key={[rowIndex, columnIndex]} row={rowIndex} column={columnIndex} symbol={val} 
                    userSymbol={userSymbol} turn={turn} makeTurn={makeTurn} />);
            });
        });
        
        return (
            <>
                <Info isPlaying={isPlaying} turn={turn}/>
                <Field>{cells}</Field>
            </>
            
        );
    }
}