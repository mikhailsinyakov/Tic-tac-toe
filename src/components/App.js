import React from 'react';
import Game from '@components/Game';
import Menu from '@components/Menu';

export default class App extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            isPlaying: false,
            userSymbol: null,
            opponentSymbol: null,
            prevResult: null
        };
        this.startGame = this.startGame.bind(this);
        this.stopGame = this.stopGame.bind(this);
    }

    startGame(userSymbol) {
        if (this.state.isPlaying) throw new Error('The game has already started');
        if (userSymbol) this.setSymbols(userSymbol);
        else this.swapSymbols();
        this.setState({ isPlaying: true });
    }

    setSymbols(userSymbol) {
        const symbols = ['×', '○'];
        if (!symbols.includes(userSymbol)) throw new Error('You should use either "×" or "○" symbol');
        const opponentSymbol = userSymbol == symbols[0] ? symbols[1] : symbols[0];
        this.setState({ userSymbol, opponentSymbol });
    }

    swapSymbols() {
        const { userSymbol, opponentSymbol } = this.state;
        if (!userSymbol || !opponentSymbol) throw new Error('You can swap symbols only if they have been set earlier');
        this.setState({ userSymbol: opponentSymbol, opponentSymbol: userSymbol });
    }

    stopGame(result) {
        if (this.state.isPlaying == false) throw new Error('The game is not playing yet');
        const possibleResults = ['Ничья', 'Вы выиграли!', 'Вы проиграли :('];
        if (!possibleResults.includes(result)) throw new Error('You should use one of the possible results');
        this.setState({ isPlaying: false, prevResult: result });
    }

    render() {
        const { startGame, stopGame, setSymbols, swapSymbols } = this;
        const { isPlaying, userSymbol, opponentSymbol, prevResult } = this.state;
        return (
            <>
                <Game stopGame={stopGame} isPlaying={isPlaying}
                        userSymbol={userSymbol} opponentSymbol={opponentSymbol} />
                <Menu isPlaying={isPlaying} prevResult={prevResult} startGame={startGame} />
            </>
        );
    }
}