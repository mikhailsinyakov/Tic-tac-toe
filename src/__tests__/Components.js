import React from 'react';
import * as TestRenderer from 'react-test-renderer';
import App from '../components/App';
import Game from '../components/Game';

let testRenderer, testInstance, instance;
beforeEach(() => {
    testRenderer = TestRenderer.create(<App/>);
    testInstance = testRenderer.root;
    instance = testInstance.instance;
});

test('it matches the snapshot', () => {
    expect(testRenderer.toJSON()).toMatchSnapshot();
});

test('when mounted should show menu and field elements and 9 cells', () => {
    const menuElement = testInstance.findByProps({id: 'menu'});
    const fieldElement = testInstance.findByProps({id: 'field'});
    const cellElements = testInstance.findAllByProps({ className: 'cell' });

    expect(menuElement).toBeTruthy();
    expect(fieldElement).toBeTruthy();
    expect(cellElements).toHaveLength(9);
});

describe('startGame method', () => {
    test('when calling without parameter in first time should throw an error', () => {
        expect(() => instance.startGame()).toThrow();
    });

    test('when calling with incorrect parameter should throw an error', () => {
        expect(() => instance.startGame('incorrect')).toThrow();
    });

    test('when calling with proper parameter should not show the menu element', () => {
        instance.startGame('×');
        expect(() => testInstance.findByProps({id: 'menu'})).toThrow();
    });

    test('when calling when the game has already started should throw an error', () => {
        instance.startGame('×');
        expect(() => instance.startGame()).toThrow();
    });

    test('the method should show the correct info message', () => {
        instance.startGame('×');
        let infoElement = testInstance.findByProps({id: 'info'});
        expect(infoElement.children).toEqual(['Ваш ход']);

        instance.stopGame('Ничья');
        instance.startGame('○');
        infoElement = testInstance.findByProps({id: 'info'});
        expect(infoElement.children).toEqual(['Ожидание хода соперника']);
    });

    test('when calling without parameter after symbols has been set should show the correct info message', () => {
        instance.startGame('×');
        instance.stopGame('Ничья');
        instance.startGame();

        const infoElement = testInstance.findByProps({id: 'info'});
        expect(infoElement.children).toEqual(['Ожидание хода соперника']);
    });

    test('when the game starts playing with "x" symbol should show 9 cells with proper style', () => {
        instance.startGame('×');
        const cellElements = testInstance.findAll(child => {
            return child.props.className == 'cell' && child.props.style.cursor == 'pointer';
        });
        expect(cellElements).toHaveLength(9);
    });

    test('when the game starts playing with "○" symbol should only show 9 cells with proper style in 2 seconds', done => {
        instance.startGame('○');
        let cellElements = testInstance.findAll(child => {
            return child.props.className == 'cell' && child.props.style.cursor == 'pointer';
        });
        expect(cellElements).toHaveLength(0);
    
        setTimeout(() => {
            cellElements = testInstance.findAll(child => {
                return child.props.className == 'cell' && child.props.style.cursor == 'pointer';
            });
            expect(cellElements).toHaveLength(8);
            done();
        }, 2000);
    });
});

describe('stopGame method', () => {
    test('if game is not playing should throw an error', () => {
        expect(() => instance.stopGame()).toThrow();
    });

    test('if parameter is not one of the possible results should throw an error', () => {
        instance.startGame('×');
        expect(() => instance.stopGame()).toThrow();
        expect(() => instance.stopGame('not possible result')).toThrow();
    });

    test('if parameter is one of the possible results should show that result in title element', () => {
        instance.startGame('×');
        instance.stopGame('Ничья');
        let titleElement = testInstance.findByProps({id: 'title'});
        expect(titleElement.children).toEqual(['Ничья']);
        
        instance.startGame('×');
        instance.stopGame('Вы выиграли!');
        titleElement = testInstance.findByProps({id: 'title'});
        expect(titleElement.children).toEqual(['Вы выиграли!']);
        
        instance.startGame('×');
        instance.stopGame('Вы проиграли :(');
        titleElement = testInstance.findByProps({id: 'title'});
        expect(titleElement.children).toEqual(['Вы проиграли :(']);
    });
});

describe('Game\'s makeTurn method', () => {
    let gameInstance;
    beforeEach(() => {
        gameInstance = testInstance.findByType(Game).instance;
    });

    test('if given not proper parameters should throw an error', () => {
        instance.startGame('×');
        expect(() => gameInstance.makeTurn(3, 2, '○')).toThrow();
        expect(() => gameInstance.makeTurn(1, 4, '○')).toThrow();
        expect(() => gameInstance.makeTurn(1, 1, '5')).toThrow();
    });

    test('if given proper parameters chosen cell should has specified symbol', () => {
        instance.startGame('×');
        gameInstance.makeTurn(1, 1, '×');
        
        const cellComponent = testInstance.findByProps({row: 1, column: 1});
        const cellElement = cellComponent.findByProps({ className: 'fas fa-times'});
        expect(cellElement).toBeTruthy();
    });

    test('if chosen cell has already symbol should throw an error', () => {
        instance.startGame('×');
        gameInstance.makeTurn(1, 1, '×');
        
        expect(() => gameInstance.makeTurn(1, 1, '○')).toThrow();
    });
});