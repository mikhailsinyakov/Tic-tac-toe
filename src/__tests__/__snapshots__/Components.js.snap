// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches the snapshot 1`] = `
Array [
  <div
    id="info"
  >
    Меню
  </div>,
  <div
    id="field"
  >
    <div
      className="cell"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    />
    <div
      className="cell"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    />
    <div
      className="cell"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    />
    <div
      className="cell"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    />
    <div
      className="cell"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    />
    <div
      className="cell"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    />
    <div
      className="cell"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    />
    <div
      className="cell"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    />
    <div
      className="cell"
      onClick={null}
      style={
        Object {
          "cursor": "default",
        }
      }
    />
  </div>,
  <div
    id="menu"
  >
    <div
      id="title"
    >
      Крестики-нолики
    </div>
    <div
      id="question"
    >
      Выберите 
      <i
        className="fas fa-times"
      />
       или 
      <i
        className="far fa-circle"
      />
    </div>
    <div
      id="options"
    >
      <i
        className="fas fa-times"
        onClick={[Function]}
      />
      <i
        className="far fa-circle"
        onClick={[Function]}
      />
    </div>
  </div>,
]
`;
